### ISSUE TITLE
SoftmaxGrad should be optimized away

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#936]
SoftmaxGrad is a generic implementation of the backprop of a gradient through Softmax. Usually, the output of softmax is fed into a cross-entropy cost, and the gradient of that cost wrt the input of softmax is much simpler than calling SoftmaxGrad.
When using mini-batches, an optimization takes care of replacing that expression with another Op, but there is no equivalent for pure stochastic gradient descent.
